v3/newrelic/instrumentation.go

```diff
@@ -36,16 +36,32 @@ func WrapHandle(app *Application, pattern string, handler http.Handler, options
 	if app == nil {
 		return pattern, handler
 	}
+
 	// add the wrapped function to the trace options as the source code reference point
-	// (to the beginning of the option list, so that the user can override this)
-	// If we came here via WrapHandleFunc, that means it will have pushed the handler's
-	// code location if possible, so the options will be (this code location), (handler function location),
-	// (other options); that results in having a backup in case either failed, with preference
-	// given to the handler's function location. In any event, the user may override all of that
-	// with a location they choose.
+	// (but only if we know we're collecting CLM for this transaction and the user didn't already
+	// specify a different code location explicitly).
+
 	return pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		options = append([]TraceOption{WithThisCodeLocation()}, options...)
-		txn := app.StartTransaction(r.Method+" "+pattern, options...)
+		var tOptions *traceOptSet
+		var txnOptionList []TraceOption
+
+		if app.app != nil && app.app.run != nil && app.app.run.Config.CodeLevelMetrics.Enabled {
+			tOptions = resolveCLMTraceOptions(options)
+			if tOptions != nil && !tOptions.SuppressCLM && (tOptions.DemandCLM || app.app.run.Config.CodeLevelMetrics.Scope == 0 || (app.app.run.Config.CodeLevelMetrics.Scope&TransactionCLM) != 0) {
+				// we are for sure collecting CLM here, so go to the trouble of collecting this code location.
+				if tOptions.LocationOverride == nil {
+					WithThisCodeLocation()(tOptions)
+				}
+			}
+		}
+		if tOptions == nil {
+			// we weren't able to curate the options above, so pass whatever we were given downstream
+			txnOptionList = options
```

Comment:
The only way to reach this case is if CLM is disabled for the whole app, in which case I think it doesn't matter whether forward the options or not. In other words I think you can just assign to `txnOptionList` up in the `if tOptions != nil && !tOptions.SuppressCLM && ...` block above and it will work out.

Reply:
True, but I'm thinking ahead to the case where there may be non-CLM TraceOptions that may be defined at some point. In that case they would still work even if we weren't doing code level metrics. We're still avoiding the unnecessary call to get the function location.

However for the time being it may be fine to leave the code simpler and expand it as appropriate at some future point where there's anything in there other than CLM options.